services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java

```diff
@@ -0,0 +1,388 @@
+                                ObjectMapper objectMapper, TimeProvider timeProvider, int preAuthorizedCodeLifeSpan) {
+        this.session = session;
+        this.bearerTokenAuthenticator = authenticator;
+        this.objectMapper = objectMapper;
+        this.timeProvider = timeProvider;
+        this.issuerDid = issuerDid;
+        this.signingServices = signingServices;
+        this.preAuthorizedCodeLifeSpan = preAuthorizedCodeLifeSpan;
+
+    }
+
+    /**
+     * Provides the URI to the OID4VCI compliant credentials offer
+     */
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("credential-offer-uri")
+    public Response getCredentialOfferURI(@QueryParam("credential_configuration_id") String vcId) {
+
+        AuthenticatedClientSessionModel clientSession = getAuthenticatedClientSession();
+
+        Map<String, SupportedCredentialConfiguration> credentialsMap = OID4VCIssuerWellKnownProvider.getSupportedCredentials(session);
+
+        LOGGER.debugf("Get an offer for %s", vcId);
+        if (!credentialsMap.containsKey(vcId)) {
+            LOGGER.debugf("No credential with id %s exists.", vcId);
+            LOGGER.debugf("Supported credentials are %s.", credentialsMap);
+            throw new BadRequestException(getErrorResponse(ErrorType.INVALID_REQUEST));
+        }
+        SupportedCredentialConfiguration supportedCredentialConfiguration = credentialsMap.get(vcId);
+        Format format = supportedCredentialConfiguration.getFormat();
+
+        // check that the user is allowed to get such credential
+        if (getClientsOfType(supportedCredentialConfiguration.getScope(), format).isEmpty()) {
+            LOGGER.debugf("No OID4VP-Client supporting type %s registered.", supportedCredentialConfiguration.getScope());
+            throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));
+        }
+
+        String nonce = generateNonce();
+        try {
+            clientSession.setNote(nonce, objectMapper.writeValueAsString(supportedCredentialConfiguration));
+        } catch (JsonProcessingException e) {
+            LOGGER.errorf("Could not convert Supported Credential POJO to JSON: %s", e.getMessage());
+            throw new BadRequestException(getErrorResponse(ErrorType.INVALID_REQUEST));
+        }
+
+        CredentialOfferURI credentialOfferURI = new CredentialOfferURI()
+                .setIssuer(OID4VCIssuerWellKnownProvider.getIssuer(session.getContext()) + "/protocol/" + OID4VCLoginProtocolFactory.PROTOCOL_ID + "/" + CREDENTIAL_OFFER_PATH)
+                .setNonce(nonce);
+
+        return Response.ok()
+                .entity(credentialOfferURI)
+                .build();
+
+    }
+
+    /**
+     * Provides an OID4VCI compliant credentials offer
+     */
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path(CREDENTIAL_OFFER_PATH + "{nonce}")
+    public Response getCredentialOffer(@PathParam("nonce") String nonce) {
```

Comment:
The specification of oid4vci dose not specify credential offer endpoint's interface.
Therefore, is it keycloak's specific interface ?

Reply:
There is a non-normative example, that I used for that https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-4.1.3 
So, yes it probably has to be seen as keycloak specific

Reply:
Thank you for the clarification.